### PR TITLE
Refactor example generation to consume server expression

### PR DESCRIPTION
**Pull request by nitinmohan87**
This PR modifies the server and CLI flags:
1. `host` flag to specify one of the hosts defined in Server expression
2. URL variables defined in Server expression are added as flags

On server side example
1. `domain` flag is added to override the domain specified in Server expression
2. `http-port` flag to specify port on which the HTTP port will listen overriding the port specified in design.

### CODE REVIEW

File: codegen/server/example_cli.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+
+	"goa.design/goa/codegen"
+	"goa.design/goa/expr"
+)
+
+// ExampleCLIFiles returns example client tool main implementation for each
+// server expression in the design.
```
Comment (raphael) [inline codegen/server/example_cli.go]: Can you remind me why we decided to generate multiple main - one per server - vs. having multiple command line flags? might be good to put that explanation in the comments too.

File: codegen/service/example_svc.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	"path/filepath"
+	"strings"
+
+	"goa.design/goa/codegen"
+	"goa.design/goa/expr"
+)
+
+type (
+	// basicEndpointData contains the data needed to render a basic endpoint
+	// implementation in the exanoke service file.
```
Comment (raphael) [inline codegen/service/example_svc.go]: typo

File: codegen/generator/example.go [hunk 3]
```diff
@@ -18,21 +19,37 @@ func Example(genpkg string, roots []eval.Root) ([]*codegen.File, error) {
 			continue // could be a plugin root expression
 		}
 
-		// Auth
-		f := service.AuthFuncsFile(genpkg, r)
-		if f != nil {
+		// example service implementation
+		if fs := service.ExampleServiceFiles(genpkg, r); len(fs) != 0 {
```
[the remark this answers is not in the document]
Comment (nitinmohan87) [reply codegen/generator/example.go]: I have changed these `Example..` functions to skip `nil` files. So we are either left with an empty slice or valid example file.

File: codegen/service/example_svc.go (hunk 2 above)
Comment (nitinmohan87) [reply codegen/service/example_svc.go]: Ha ha.. That's really bad one.

File: codegen/server/example_cli.go (hunk 1 above)
Comment (nitinmohan87) [reply codegen/server/example_cli.go]: Wasn't that your original intention behind using the server expression in generating example mains? One for each server expression. The hosts and URI variables within each server expression will get their own CLI args.

Comment (raphael) [reply codegen/server/example_cli.go]: Duh yeah you are right. One server generated per server expression :)

**Comment by difengo** [comment]
In the "example_scv.go" file, a specific struct named "basicEndpointData" is used to hold the data. But this struct is private. This creates an issue for plugins that need to read the values from the data before modifying the generated endpoint. Would it be possible to make it public instead ?

**Pull request merged**